v22.1/security-reference/scram-authentication.md

```diff
@@ -0,0 +1,323 @@
+
+In the CockroachDB SQL shell, use the computed hash as the value of `PASSWORD` when [creating a user](../create-user.html) or [rotating the passwors for an existing user](../alter-user.html).
+
+{% include_cached copy-clipboard.html %}
+```sql
+CREATE USER cool_user WITH PASSWORD 'SCRAM-SHA-256$119680:D1nWfUQfRP2HWbVuK19Ntw==$ulRLI4fp/bvJN/oBC3x6Q8daLnPit2eWLvm2sex9vyc=:XHVeRmT8iboSn0KMbIx3nMRrebR2gIbi+qEMnHfJi+I=';
+
+ALTER USER other_user WITH PASSWORD "$scram$100000$vZcyRiil1BojRGhN$sha-1=pRLECKY8rLwZKkmhKu02b8RfIeQ,sha-256=ClkmygvxmE9tOx8swe2lD5kb/4rjE3qa2BF6mAhshHE,sha-512=l3zLEbJLHqRZGi5uCdwDqkNzk4bfoZTcvyKRXyu6YP3mtoh7JFclxKdAKXKQTBAWMQDv.pbNgUt3/KOX4Hgteg";
+```
+```
+CREATE ROLE
+
+Time: 156ms total (execution 156ms / network 0ms)
+```
+
+### Inspect the user
+
+In this way, the server stores the information needed for SCRAM authentication, despite the plaintext password never having been conveyed to CockroachDB. The 'concern' with plaintext passwords has been cleanly 'separated' from CockroachDB.
+
+{% include_cached copy-clipboard.html %}
+```sql
+SELECT username, "hashedPassword" FROM system.users WHERE username='cool_user'; 
+SELECT *  FROM system.users WHERE username = "cool_user";
+```
+```
+  username  |                                                             hashedPassword
+------------+------------------------------------------------------------------------------------------------------------------------------------------
+  cool_user | SCRAM-SHA-256$119680:6XN0y7A83hylmX+3uMRPvQ==$mOI18rsucBjSrNMDDqNJ1/q4cROvRDLIUqjC0BzQHEg=:4zPCVFflWWHD53ZEIrcH4q3X1+le86TBIV0Dce6fIuc=
+```
+
+## Appendix: Python SCRAM-hashing script
+
+```python
+# cockroach-scram-hasher.py
+
+from base64 import standard_b64encode
+from hashlib import pbkdf2_hmac, sha256
+from os import urandom
+import hmac
+import sys
+
+salt_size = 16
+digest_len = 32
+iterations = 4096
+
+def b64enc(b: bytes) -> str:
+    return standard_b64encode(b).decode('utf8')
+
+def pg_scram_sha256(passwd: str) -> str:
+    salt = urandom(salt_size)
+    digest_key = pbkdf2_hmac('sha256', passwd.encode('utf8'), salt, iterations,
+                             digest_len)
+    client_key = hmac.digest(digest_key, 'Client Key'.encode('utf8'), 'sha256')
+    stored_key = sha256(client_key).digest()
+    server_key = hmac.digest(digest_key, 'Server Key'.encode('utf8'), 'sha256')
+    return (
+        f'SCRAM-SHA-256${iterations}:{b64enc(salt)}'
+        f'${b64enc(stored_key)}:{b64enc(server_key)}'
+    )
+
+def print_usage():
+    print("Usage: provide single password argument to encrypt")
+    sys.exit(1)
+
+def main():
+    args = sys.argv[1:]
+
+    if args and len(args) > 1:
+        print_usage()
+
+    if args:
+        passwd = args[0]
+    else:
+        passwd = sys.stdin.read().strip()
+
+    if not passwd:
+        print_usage()
+
+    print(pg_scram_sha256(passwd))
+
+if __name__ == "__main__":
+    main()
+```
+
+## References
```

Comment:
Another possible add to the style guide.... adding references to external sites in a separate reference section. All good things to consider.